### PR TITLE
Quick Fix: Small issues in profiles

### PR DESCRIPTION
**Pull request by jellypuno**
## Proposed changes

These are small changes and are not high priority.

Changes Include:
- Remove the word Optional in Username and Password
- Display the error message when Profile Inactive is encountered
- Handle the additional fields when creating and editing a profile when the user cancelled it.

## Release Notes

Milestone:

Changelog:

## Types of changes

What types of changes does your code introduce to Zowe Explorer?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Updates to Documentation or Tests (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This checklist will be used as reference for both the contributor and the reviewer_

- [x] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/vscode-extension-for-zowe/wiki/Best-Practices:-Contributor-Guidance) wiki
- [x] PR title follows [Conventional Commits Guidelines](https://www.conventionalcommits.org/en/v1.0.0-beta.2/)
- [x] PR Description is included
- [ ] gif or screenshot is included if visual changes are made
- [x] `npm run vscode:prepublish` has been executed
- [x] All checks have passed (DCO, Jenkins and Code Coverage)
- [x] I have added unit test and it is passing
- [ ] I have added integration test and it is passing
- [x] There is coverage for the code that I have added
- [x] I have tested it manually and there are no regressions found
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any PR dependencies have been merged and published (if appropriate)

## Further comments

If the validation is on and the profile that you are using doesn't have a username and password or token, you will encounter this error:
image

I have opened an issue in CLI for this:
https://github.com/zowe/zowe-cli/issues/875

### CODE REVIEW

**Review by JillieBeanSim — COMMENTED** [review]
@jellypuno this is looking good but I did get this error when updating a profile and going all the way through the update process.  It fully updated the profile, yaml included but got the error on top of successful message
image

**Comment by jellypuno** [comment]
> @jellypuno this is looking good but I did get this error when updating a profile and going all the way through the update process. It fully updated the profile, yaml included but got the error on top of successful message
> image

Thank you for testing @JillieBeanSim . I will take a look at this tomorrow.

File: packages/zowe-explorer/src/Profiles.ts [hunk 1]
```diff
@@ -482,10 +482,24 @@ export class Profiles extends ProfilesCache {
                             if (!Number.isNaN(Number(updValue))) {
                                 updSchemaValues[value] = Number(updValue);
```
Comment (zFernand0) [inline packages/zowe-explorer/src/Profiles.ts:483]: If input is an empty string (i.e. the user hit enter), `Number("") === 0`
This results in the profile still being saved with `responseTimeout: 0`.

We could check for `updValue !== "" && !Number.isNaN(Number(updValue)` but that leads to another problem. (see next comment)

This comments applies to both Creating a new Connection (`createNewConnection`) and Editing a session (`editSession`)

File: packages/zowe-explorer/src/Profiles.ts [hunk 2]
```diff
@@ -482,10 +482,24 @@ export class Profiles extends ProfilesCache {
...
-                                } else {
-                                    break;
+                                switch (true) {
+                                    case schema[value].optionDefinition.hasOwnProperty("defaultValue"):
+                                        if (schema[value].optionDefinition.defaultValue === null) {
+                                            vscode.window.showInformationMessage(
+                                                localize("editConnection.rejectUnauthorize", "Operation Cancelled")
+                                            );
+                                            return undefined;
+                                        }
```
Comment (zFernand0) [inline packages/zowe-explorer/src/Profiles.ts:492]: If the user didn't enter a value for a numeric property, and the schema contains a null value, we should still create the profile without this property.

For comparison, here are the schema definitions encoding and responseTimeout
```
{
  "name": "encoding",
  "aliases": ["ec"], 
  "description": "The encoding for download and upload of z/OS data set and USS files. The default encoding if not specified is 1047.",
  "type":"number"
}
```

```
{
  "name": "responseTimeout",
  "aliases": ["rto"],
  "description": "The maximum amount of time in seconds the z/OSMF Files TSO servlet should run before returning a response. Any request exceeding this amount of time will be terminated and return an error. Allowed values: 5 - 600",
  "type": "number",
  "defaultValue": null,
  "numericValueRange": [5,600]
}
```

**Review by zFernand0 — COMMENTED** [review]
Added a couple of comments 😋

**Comment by jellypuno** [comment]
@zFernand0 Just a clarification, are you saying that if the user enters a blank in the property (e.g. responseTimeOut) we should create the profile with a blank value instead of zero? or are saying that we should not add the property into the profile?

**Comment by jellypuno** [comment]
Since this is a generic checking of number, If I try to allow the creation of profile that is equal to null, we are having this error.

image

In my opinion, if it is defined as a number, the default number should not be null. I opened an issue for that: https://github.com/zowe/zowe-cli/issues/871

**Comment by zFernand0** [comment]
> are you saying that if the user enters a blank in the property (e.g. responseTimeOut) we should create the profile with a blank value instead of zero? or are saying that we should not add the property into the profile?

The latter. We should not store properties that: 
  - Do not have a default value set by the schema
  - The user didn't specify a value

**Comment by jellypuno** [comment]
@zFernand0 I plan to add this in the future. The goal of this PR is just to handle those additional fields. Because right now, if I cancelled the creation / editing with those additional fields it will display an error and not "Operation Cancelled"

**Comment by katelynienaber** [comment]
@jellypuno 2 things:

1: Does this fix [this issue (#1065)](https://github.com/zowe/vscode-extension-for-zowe/issues/1065)?
2: I think it might be better to store empty strings than totally remove the fields. The reason is that Imperative is using `deepmerge` to merge old and new profiles when editing them, and if the new profile doesn't have all the same fields as the old one, `deepmerge` retains the old fields (and their values) when merging. See [this PR](https://github.com/zowe/vscode-extension-for-zowe/pull/1120) for some affected fields...maybe there is a better way to deal with this, I am not sure.

**Comment by jellypuno** [comment]
> @jellypuno 2 things:
> 
> 1: Does this fix [this issue (#1065)](https://github.com/zowe/vscode-extension-for-zowe/issues/1065)?
> 2: I think it might be better to store empty strings than totally remove the fields. The reason is that Imperative is using `deepmerge` library to merge a profile's old and new values when updating the profile. If the new profile doesn't have all the same fields as the old one, `deepmerge` retains the old fields (and their values) when merging. See [this PR](https://github.com/zowe/vscode-extension-for-zowe/pull/1120) for some fields affected by `deepmerge`. Maybe there is a better way to deal with this than storing empty strings...what do you think?

I don't think I'm removing it. I'm storing it as spaces or default value. The handling that I added is that if you cancelled the Profile Creation on those new fields the error message is better. But maybe I need to double check

**Comment by katelynienaber** [comment]
> > @jellypuno 2 things:
> > 1: Does this fix [this issue (#1065)](https://github.com/zowe/vscode-extension-for-zowe/issues/1065)?
> > 2: I think it might be better to store empty strings than totally remove the fields. The reason is that Imperative is using `deepmerge` library to merge a profile's old and new values when updating the profile. If the new profile doesn't have all the same fields as the old one, `deepmerge` retains the old fields (and their values) when merging. See [this PR](https://github.com/zowe/vscode-extension-for-zowe/pull/1120) for some fields affected by `deepmerge`. Maybe there is a better way to deal with this than storing empty strings...what do you think?
> 
> I don't think I'm removing it. I'm storing it as spaces or default value. The handling that I added is that if you cancelled the Profile Creation on those new fields the error message is better. But maybe I need to double check

I was referring to your conversation with @zFernand0 ...so I will tag him in that comment :)

**Review by katelynienaber — CHANGES_REQUESTED** [review]
Looks good and works as expected. The basePath is still not updated when the profile is edited...but I think that's out of scope and is already fixed in [another PR](https://github.com/zowe/vscode-extension-for-zowe/pull/1120).

IMHO though we should maybe leave the `optional` placeholder when creating/editing a profile, but remove it when prompted for username/password.

**Comment by jellypuno** [comment]
> Looks good and works as expected. The basePath is still not updated when the profile is edited...but I think that's out of scope and is already fixed in [another PR](https://github.com/zowe/vscode-extension-for-zowe/pull/1120).
> 
> IMHO though we should maybe leave the `optional` placeholder when creating/editing a profile, but remove it when prompted for username/password.

I removed the optional word because it is used by ssoLogin and it is not supposed to optional.

**Comment by katelynienaber** [comment]
> > Looks good and works as expected. The basePath is still not updated when the profile is edited...but I think that's out of scope and is already fixed in [another PR](https://github.com/zowe/vscode-extension-for-zowe/pull/1120).
> > IMHO though we should maybe leave the `optional` placeholder when creating/editing a profile, but remove it when prompted for username/password.
> 
> I removed the optional word because it is used by ssoLogin and it is not supposed to optional. It didn't make sense to create another text just for ssoLogin so I just removed the optional because we intend to remove the Optional Credentials in the future

Ah okay, SO GLAD we will remove optional credentials! I didn't know that was the plan so I will approve now.

**Review by katelynienaber — APPROVED** [review]
LGTM!

**Review by JillieBeanSim — COMMENTED** [review]
Hey @jellypuno when I was creating and editing a zosmf profile, I hit the escape at the encoding value.  I was expecting the `operation cancelled` message to pop up then but instead it went to the next textbox for response timeout then I clicked escape again to get the `operation cancelled` message.
Is this expected, as I was not expecting to get prompted with another textbox after hitting escape the first time?

**Comment by jellypuno** [comment]
> Hey @jellypuno when I was creating and editing a zosmf profile, I hit the escape at the encoding value. I was expecting the `operation cancelled` message to pop up then but instead it went to the next textbox for response timeout then I clicked escape again to get the `operation cancelled` message.
> Is this expected, as I was not expecting to get prompted with another textbox after hitting escape the first time?

Did you get the latest commit? It is not happening for me.

operCancel

**Review by JillieBeanSim — DISMISSED** [review]
@jellypuno I must have been missing a commit.  This LGTM, functionality is working as expected but I do get a few failing integration tests that deal with creating a profile now that I didn't get on Master branch. 

image

**Comment by jellypuno** [comment]
@JillieBeanSim I fixed the integration test. Thank you for testing and reviewing this PR.

**Review by katelynienaber — APPROVED** [review]
LGTM :)

**Review by JillieBeanSim — APPROVED** [review]
This LGTM, all expected tests and checks are passing and functionality works great! Thanks for all of your work on this @jellypuno, fixing these small issues make a huge positive impact for users.

**Pull request merged**